analyzer/validatorstakinghistory/validatorstakinghistory.go

```diff
@@ -0,0 +1,193 @@
+
+	"github.com/oasisprotocol/nexus/analyzer"
+	"github.com/oasisprotocol/nexus/analyzer/item"
+	"github.com/oasisprotocol/nexus/analyzer/queries"
+	"github.com/oasisprotocol/nexus/config"
+	staking "github.com/oasisprotocol/nexus/coreapi/v22.2.11/staking/api"
+	"github.com/oasisprotocol/nexus/log"
+	"github.com/oasisprotocol/nexus/storage"
+	"github.com/oasisprotocol/nexus/storage/oasis/nodeapi"
+)
+
+// The validator staking history analyzer (1) gets the next epoch / validators to download validator
+// history for, (2) downloads that info, and (3) saves the info in the database.
+//
+// We process epochs sequentially in chronological order in order to track validator
+// history for all past and current validators. More concretly, for a given epoch, we find all current
+// validators and take the union with the set of all tracked validators from the previous epoch.
+// Notably, we do not track validator history for an entity before it first becomes an active validator.
+//
+// WARNING: This analyzer SHOULD NOT run while block analyzers are in fast sync. This analyzer expects that
+// when a row appears in `chain.epochs`, its first block is already processed. In fast sync, blocks are processed
+// out of order. (As of 2024-06, fast-sync actually does not update chain.epochs directly so it might not cause bugs
+// when running in parallel with this analyzer, but that's by chance rather than by design.)
+
+const (
+	validatorHistoryAnalyzerName = "validator_history"
+)
+
+type processor struct {
+	source     nodeapi.ConsensusApiLite
+	target     storage.TargetStorage
+	startEpoch uint64
+	logger     *log.Logger
+}
+
+var _ item.ItemProcessor[*Epoch] = (*processor)(nil)
+
+func NewAnalyzer(
+	initCtx context.Context,
+	cfg config.ItemBasedAnalyzerConfig,
+	startHeight uint64,
+	sourceClient nodeapi.ConsensusApiLite,
+	target storage.TargetStorage,
+	logger *log.Logger,
+) (analyzer.Analyzer, error) {
+	logger = logger.With("analyzer", validatorHistoryAnalyzerName)
+
+	// Find the epoch corresponding to startHeight.
+	if startHeight > math.MaxInt64 {
+		return nil, fmt.Errorf("startHeight %d is too large", startHeight)
+	}
+	epoch, err := sourceClient.GetEpoch(initCtx, int64(startHeight))
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch epoch for startHeight %d", startHeight)
+	}
+	p := &processor{
+		source:     sourceClient,
+		target:     target,
+		startEpoch: uint64(epoch),
+		logger:     logger,
+	}
+
+	return item.NewAnalyzer[*Epoch](
+		validatorHistoryAnalyzerName,
+		cfg,
+		p,
+		target,
+		logger,
+	)
+}
+
+type Epoch struct {
+	epoch          uint64
+	startHeight    int64
+	prevValidators []signature.PublicKey // Validator set from the previous epoch.
+}
+
+// Note: limit is ignored here because epochs must be processed sequentially in chronological order.
+func (p *processor) GetItems(ctx context.Context, limit uint64) ([]*Epoch, error) {
+	var epochs []*Epoch
+	rows, err := p.target.Query(
```

Comment:
nit: Could use `QueryRow` since we query a single row?

Reply:
Good point, it ends up being cleaner that way. Updated!